### ISSUE TITLE
False negative type mismatch with associated types in arguments

### ISSUE THREAD

**Issue by HKalbasi** [rust-lang/rust-analyzer#15321]
```Rust
trait A {
    fn m(&self) -> i32;
}

fn x<T: B>(k: &<T as B>::Ty) {
    k.m();
}

impl A for u8 {
    fn m(&self) -> i32 {
        8
    }
}

impl A for u32 {
    fn m(&self) -> i32 {
        32
    }
}

trait B {
    type Ty: A;
}

impl B for u16 {
    type Ty = u32;
}

#[test]
fn ttt() {
    x::<u16>(&2u8); // Expected u32 found u8 is missing here.
}
```
